modules/data.atmosphere/R/downscale.met.R

```diff
@@ -0,0 +1,425 @@
+                   solarnoon=as.numeric(format(solar_noon$time, "%H%M")),
+                   sunset=as.numeric(format(sunset$time, "%H%M")),
+                   day_length=as.numeric(sunset$time-sunrise$time)) 
+        
+      }
+      
+      swmodel <- model$surface_downwelling_shortwave_flux_in_air
+      swdn <- vector()
+    
+      # The sine method produces an hourly sine wave of 
+      if (swdn_method == "sine"){
+        
+        eph <- ephemeris(lat_flux,lon_flux,date = paste0(eph_year,"-01-01 00:00:00"), 
+                       span = sp, tz = "UTC")
+        day_len <- eph$day_length
+        
+        # Need to have average daily values for this method, so this upscales the model data to daily resolution if needed
+        daily_row = nrow(model)
+        daily_step = daily_row/sp
+        daily.swdn = vector()
+          for (x in 1:sp){
+            daily.swdn[x] <- mean(swmodel[(x*daily_step-daily_step+1):(x*daily_step)])}
+        
+        # creating the sine wave
+        for (i in seq_along(daily.swdn)){
+          t <- seq(from=pi/day_len[i],to=pi,by=pi/day_len[i])
+          wav <- ((daily.swdn[i]*(24/day_len[i]))/0.637)*sin(t)
+          
+        # swdn = 0 without sunlight
+        srs <- eph$sunrise
+        hr <- substr(srs[i],1,2)
+        hr <-as.numeric(hr)
+        hr <- hr+utc_diff
+        
+        l <- vector()
+        for (n in seq_along(1:hr)){
+          l[n] = 0}
+        for (n in seq_along(wav)){
+          l[n+hr] = wav[n]}
+        for(n in seq_along(1:(24-(length(wav)+hr)))){
+          l[n+hr+length(wav)] = 0}
+        
+        swdn <- append(swdn,l)}
+        
+        swflux <- vector()
+        sw_step <- length(swdn)/reso_len
+        for (x in 1:reso_len){
+          swflux[x] <- mean(swdn[(x*sw_step-sw_step+1):(x*sw_step)])}
+        swflux[swflux < 0] = 0 
+      }
+        
+      # The spline method uses spline interpolation to connect existing values and downscale
+      if (swdn_method == "spline"){
+        tem.met = vector()
+          for (x in seq(from=0, to=nrow(upscale_flux), by=div)){
+            tem.met[x] = swmodel[x/div]}
+        
+      swflux = vector()
+      swflux = zoo::na.spline(tem.met)
+      swflux[swflux<0] <- 0
+      }
+      
+      # The Waichler method doesn't need averaged SW flux values, it models SW downwelling flux based on Tmax-Tmin and Precipitation
+      # Reference is Waichler and Wigtosa 2003. Our no-precip coefficient is 2 instead of 1 (fits the data better)
+      if (swdn_method == "Waichler"){
+        inter <- paste0(reso," hour")
+        days <- seq(as.POSIXct(paste0(eph_year,"-01-01 00:00:00", tz = "US/Central")), 
+                    as.POSIXct(paste0(eph_year,"-12-31 18:00:00", tz = "US/Central")), by=inter)
+        
+        Z <- Ratmosphere::SZA(days,lat_flux, lon_flux)
+        I <- 1000*cos_d(Z)
+        m <- vector()
+        for (i in seq_along(1:12)){
+          m[i] <- Hmisc::monthDays(as.Date(paste0(year,'-',i,'-01')))}
```

Comment:
Similar to `rnorm` comment above - `Hmisc::monthdays` is parallelized, so why do this in a loop?